src/backend/distributed/README.md

```diff
@@ -0,0 +1,2472 @@
+
+- **Structure and Source**:
+  The `combineQuery` can be traced back through the `DistributedPlan->combineQuery` struct. This query is essentially constructed in the `CreatePhysicalDistributedPlan` function. However, the actual source comes from `MasterExtendedOpNode()` within the logical optimizer. For deeper insights into this logic, you can refer to the paper and video links shared under the "Logical Planner & Optimizer" section.
+
+- **Example**:
+  The simplest example is the following where Citus sends `count(*)` to the shards, and needs to do a `sum()` on top of the results collected from the workers.
+  ```sql
+  SET client_min_messages TO DEBUG4;
+  DEBUG:  generated sql query for task 1
+  DETAIL:  query string: "SELECT count(*) AS count FROM public.users_table_102008 users_table WHERE true"
+  ....
+  DEBUG:  combine query: SELECT COALESCE((pg_catalog.sum(count))::bigint, '0'::bigint) AS count FROM pg_catalog.citus_extradata_container(10, NULL::cstring(0), NULL::cstring(0), '(i 1)'::cstring(0)) remote_scan(count bigint)
+  D
+  ```
+
+
+### CTE Processing
+
+- **In Postgres 13 and Later Versions**:
+In Postgres 13 and later versions, CTEs (Common Table Expressions) are almost like subqueries. Usually, these CTEs are transformed into subqueries during `standart_planning()`. Citus follows the same approach via `RecursivelyInlineCtesInQueryTree()`.
+
+- **Additional Consideration in Citus**:
+For Citus, there's an additional consideration. CTEs that aren't inlined get materialized. In the Citus context, materialization converts these CTEs into intermediate results. Some users leverage this for achieving full-SQL coverage.
+
+- **Extra CTE Check in Citus**:
+ Citus includes an extra check before inlining CTEs, conducted by the function `TryCreateDistributedPlannedStmt`. Here, the planner first tries to inline all CTEs and then checks whether Citus can still plan the query. If not, the CTEs remain as is, leading to their materialization. If all CTEs are materialized (e.g., read_intermediate_result), then the query becomes equivalent of a query on reference table, hence full SQL.
+
+ **Examples for Better Understanding**:
+ I understand the logic might seem complex at first. Simple examples will be provided for better understanding.
+
+```sql
+-- a CTE that is inlined as subquery, and does a query-pushdown
+WITH cte_1 AS (SELECT DISTINCT user_id FROM orders_table)
+SELECT * FROM cte_1;
+
+```
+
+So, from Citus' query planning perspective
+ the above CTE is equivalent to the following subquery
+
+ ```sql
+SELECT * FROM
+	(SELECT DISTINCT user_id FROM orders_table) cte_1;
+```
+
+Once a CTE is inlined, then the rest of the query
+ planning logic kicks in
+for example, below, the cte is inlined and then
+because the subquery is NOT safe to pushdown
+it is recurively planned
+```sql
+WITH cte_1 AS (SELECT DISTINCT product_id FROM orders_table)
```

Comment:
Here the example does `DISTINCT product_id`, which turns out to be NOT safe to pushdown.

However, the two SQL code blocks above it are using `DISTINCT user_id` (which is the distribution key of the orders_table). Would that make it safe to pushdown?

Reply:
Yes, but the example tries to show a case where after inlining the CTE, we could end up with a non-push downable subquery?

Reply:
Yes, I understood.